[Networking] [E2E]  n-to-1 iperf client->server throughput benchmarking

jayunit100 opened the pull request.

jayunit100 commented:
@kubernetes/sig-testing a first pass ateempt at what we discussed tuesday, using iperf, which measures netowrk performance. 
This is the first PR for #22707

jayunit100 commented:
I've now run this on GCE several times and its stable.
Ready for a first look from anyone interested.
I'll remove the `DO-NOT-MERGE` after taking a fresh look at this in the morning.
```
[begin] Node,Bandwith CSV
iperf-e2e-cli-pod-0-10.248.207.5-3--server=10.0.206.216 	  1.595592E+10
iperf-e2e-cli-pod-1-10.251.70.5-3--server=10.0.206.216 	  1.678759E+09
iperf-e2e-cli-pod-2-10.245.55.5-3--server=10.0.206.216 	  1.484329E+09
[end] Node,Bandwith CSV
```

jayunit100 commented:
I've now run this on GCE several times and its stable.
Ready for a first look from anyone interested.
I'll remove the DO-NOT-MERGE after taking a fresh look at this in the morning.

A Test result for 3 nodes follows -- this is done on GCE.

[begin] Node,Bandwith CSV
iperf-e2e-cli-pod-0-10.248.207.5-3--server=10.0.206.216       1.595592E+10
iperf-e2e-cli-pod-1-10.251.70.5-3--server=10.0.206.216    1.678759E+09
iperf-e2e-cli-pod-2-10.245.55.5-3--server=10.0.206.216    1.484329E+09
[end] Node,Bandwith CSV

jayunit100 commented:
Removed the do-not-merge label, and ran a few tests of this on GCE.  It stable from what I can tell, but I labelled as flakey since it is a new test which we will want to bake into the CI over time.

timothysc commented on test/e2e/networking_perf.go:
```diff
@@ -0,0 +1,238 @@
+)
+
+// Declared as Flakey since it has not been proven to run in parallel on small nodes or slow networks in CI
+var _ = Describe("Networking IPerf [Experimental] [Networking] [Flaky]", func() {
```
Please remove [Flaky]

jayunit100 replied:
to be clear, this test isn't flaky by definition, but it has the potential to be in an unstable cloud, and we dont wanna destabilize the baseline CI at this time when so much is going on. hence the label.

timothysc replied:
KDescribe ;-)

timothysc commented on test/e2e/networking_perf.go:
```diff
@@ -0,0 +1,238 @@
+
+// Declared as Flakey since it has not been proven to run in parallel on small nodes or slow networks in CI
+var _ = Describe("Networking IPerf [Experimental] [Networking] [Flaky]", func() {
+	gceBandwidthBits := int64(30000000000)
```
Magic numbers should be constants.

timothysc commented on test/e2e/networking_perf.go:
```diff
@@ -0,0 +1,238 @@
+	run(f, 1, 10000, gceBandwidthBits)
+})
+
+func run(f *Framework, numServer int, numClient int, maxBandwidthBits int64) {
```
Given that there may be multiple tests in this file you may want to call this something less generic.

timothysc commented on test/e2e/networking_perf.go:
```diff
@@ -0,0 +1,238 @@
+			appName,
+			api.Container{
+				Name:  "iperf-server",
+				Image: "mattgruter/iperf",
```
Given that iperf doesn't support numServer, you may want to either change the signature or factor out the images.

jayunit100 replied:
agree ; `numServer` should be  a const until / unless i scrape endpoints and run iperf that way (which is possible in the next iteration)

timothysc replied:
@ixdy what's upstream policy about using 3rd party containers for e2es.  Typically they are all google owned.

ixdy replied:
we've had flakiness issues pulling from Docker hub, yes, so we prefer anything merge-blocking to be on gcr.io.

timothysc commented on test/e2e/networking_perf.go:
```diff
@@ -0,0 +1,238 @@
+		expectedBandwidth := int(float64(maxBandwidthBits) / float64(totalPods))
+		Expect(totalPods).NotTo(Equal(0))
+		appName := "iperf-e2e"
+		CreateServiceWithPods(f,
```
In general I wish there were a single location for our testing image specs that we use.  ~= yaml injection.

timothysc commented on test/e2e/networking_perf.go:
```diff
@@ -0,0 +1,238 @@
+			// this will be used to generate the -service name which all iperf clients point at.
+			numServer, // 1 server
+		)
+		CreateServiceWithPods(f,
```
Aren't you just creating a wide-rc here?  why not call it such?

jayunit100 replied:
we're manually writing a pod to each node, for up to N nodes max, so its "sort of" like a replication controller but not really.

timothysc commented on test/e2e/networking_perf.go:
```diff
@@ -0,0 +1,238 @@
+			numClient, // 1 server
+		)
+		Logf("Reading all perf results to stdout.")
+		Logf("date,cli,cliPort,server,serverPort,id,interval,transferBits,bandwidthBits")
```
formatting on this line appears weird in github.. could be github.

jayunit100 commented:
comments addressed

jayunit100 commented on test/e2e/networking_perf.go:
```diff
@@ -0,0 +1,249 @@
+func runClientServerBandwidthMeasurement(f *Framework, numClient int, maxBandwidthBits int64) {
+
+	// TODO: Make this a function parameter, once we distribute iperf endpoints.
+	numServer := 1
```
i didnt make this  a constant because this really should variable... but for now, its simpler to hardcode it since numServer parameterization will be a follow on feature

jayunit100 committed:
E2E Network performance run which uses iperf.
generic pod-per-node functionality for testing - 2 node test only
- update framework to decompose pod vs svc creation for composition.
- remove hard coded 2 and pointer to --scale

spxtr commented:
Would someone from @kubernetes/sig-network take a look at this?

bprashanth commented:
I can try and take a look sometime tomorrow unless someone else jumps on this. We're no longer using the e2e tags you have now, like [flaky]. See https://github.com/kubernetes/kubernetes/blob/8cc91944e0555af82f86cc7f21fb638be60665fd/docs/devel/e2e-tests.md#kinds-of-tests

bprashanth commented:
so this would probably be [feature:network-perf] or something (pick one closest to the pattern established by existin feature: perf tests)

jayunit100 commented:
Updated the flakey label to "Feature:Network-Performance".

timothysc commented on test/e2e/networking_perf.go:
```diff
@@ -0,0 +1,249 @@
+		}()
+
+		start := time.Now()
+		for len(bandwidthMap) < expectedCli && time.Since(start) <= 2*time.Minute {
```
Please put the 2 minute time constant as a const above.

jayunit100 replied:
Hmmm... Not sure if this should be a constant.  I think really we should scale it w/respect to client number.... but point taken - it should either be a constant or it should be dynamic./

timothysc commented on test/e2e/util.go:
```diff
@@ -2298,6 +2298,20 @@ func ListSchedulableNodesOrDie(c *client.Client) *api.NodeList {
 	return nodes
 }
 
+func ListReadyNodes(f *Framework) (nodes *api.NodeList, err error) {
```
comment exported f(n)s

timothysc commented on test/e2e/util.go:
```diff
@@ -3753,3 +3767,29 @@ func CoreDump(dir string) {
+// Meant to be called in a loop - where the fn does some assertion logic on a test (i.e. look for string in the pod logs)
+func TryEachPod(c *client.Client, ns, selectorKey, selectorValue string, fn func(api.Pod)) {
+	pods := []api.Pod{}
+	for t := time.Now(); time.Since(t) < podListTimeout; time.Sleep(poll) {
```
isn't this similar to what the wait function in utils does?  we should use that one.

jayunit100 replied:
This particular function selects pods which don't have an associated RC and re run's a lambda on them.  Other functions... Its a little unique... I guess you're asking "why not wait for all pods running?"...

- Because it allows for nice realtime feedback.  Waiting for all pods running doesn't really tell us much when running on a large cluster (30,40,...,nodes).
- We also want to run this function even against pods that aren't running (i.e. PodSucceeded state). Remember that a GB of so can easily be sent quickly on a modern network.  So many iperf pods will simply complete in a few seconds, we want to continue polling them and keep updating results, not just wait till the slowest one is finished.

- I agree with you that the number of "do this a bunch of times with these pods" functions is getting astronomical.  I think there are like 25 different ones or something, in util.go.    

We definetely need to audit them for redundancy's  in  #23540, in my case at first glance i don't see anything that does exactly what i need here, but maybe i can modularize one of them.

jayunit100 replied:
UPDATE tim is referring specifically to the 3775 line rather than not the waitForPodRunning in namespace methods in test/e2e/utils.go ..... so this particular line we can replace w/ until

timothysc commented on test/e2e/util.go:
```diff
@@ -3753,3 +3767,29 @@ func CoreDump(dir string) {
+		podList, err := c.Pods(ns).List(options)
+		Expect(err).NotTo(HaveOccurred())
+		for _, pod := range podList.Items {
+			if pod.Status.Phase == api.PodPending || pod.Status.Phase == api.PodRunning || pod.Status.Phase == api.PodSucceeded {
```
This or logic on pods should probably be a separate func, or applied as the fn() operation b/c we're pre-filtering the list.

timothysc commented on test/e2e/util.go:
```diff
@@ -3753,3 +3767,29 @@ func CoreDump(dir string) {
+		}
+	}
+	if pods == nil || len(pods) == 0 {
+		Failf("No pods found")
```
Fail?  or return error so caller using the util should fail based on their test.

jayunit100 replied:
Good catch I agree, Failf is overspecified, especially for a utils function.

timothysc commented on test/e2e/networking_perf.go:
```diff
@@ -0,0 +1,249 @@
+
+		start := time.Now()
+		for len(bandwidthMap) < expectedCli && time.Since(start) <= 2*time.Minute {
+			TryEachPod(f.Client, f.Namespace.Name, "app", "iperf-e2e-cli-pod",
```
s/TryEachPod/TryEachRunningPod

jayunit100 replied:
toss-up --- This actually targets more than just running pods.  Are you sure you want `TryEachRunningSucceededPendingPod` :) ?

timothysc commented on test/e2e/networking_perf.go:
```diff
@@ -0,0 +1,249 @@
+					if err == nil {
+						Logf(resultS)
+						ipr := NewIPerf(resultS)
+						bandwidthMap[p.Name+"-"+ipr.cli+"-"+ipr.id+"--server="+ipr.server] = ipr.bandwidthBits
```
Why not create a IPerfResults struct which is this map + a util to translate the IPerfResult, as that would make is easier to read.   Plus the logic below could go into an add f(n)

jayunit100 replied:
yeah, IPerfResults good idea

timothysc commented on test/e2e/networking_perf.go:
```diff
@@ -0,0 +1,249 @@
+		}
+
+		fmt.Println("[begin] Node,Bandwith CSV")
+		for node, bandwidth := range bandwidthMap {
```
Same thing you could have a IPerfResults print or spew f(n)

jayunit100 replied:
Yup agree

timothysc commented on test/e2e/networking_perf.go:
```diff
@@ -0,0 +1,249 @@
+}
+
+// parse parses an IPerf CSV output line into an IPerfResult
+func NewIPerf(csvLine string) IPerfResult {
```
This seems rigidly tied to the version of iperf.  

Is there a more verbose output that could be more resilient to changes in iperf? e.g. name=value

jayunit100 replied:
- I don't think iperf2 supports json, and the `-o` option seemed kind of broken actually.  If this is a problem, do you think we should try iperf3?  
- My thought process was "iperf2 seems to work, and single lines are easy enough to parse, so we can just use this struct as the abstraction and update it as necessary if we use iperf3.
- i think iperf3 containers support JSON.  
- From what i could tell, iperf2 output formatting was pretty limited and there wasn't a whole lot of flexibility. 

I figured it might be simpler to start with this, and move to iperf3 w/ json later if we need it.  But I'm open to updating to iperf3 and using their json output format.

timothysc commented on test/e2e/networking_perf.go:
```diff
@@ -0,0 +1,249 @@
+
+type StrSlice []string
+
+func (s StrSlice) get(i int) string {
```
There has to be util equiv somewhere in the codebase.

timothysc commented on test/e2e/networking_perf.go:
```diff
@@ -0,0 +1,249 @@
+
+// return an int or fail.  the first arg is for easy debugging purposes only in case of failure.
+func intOrFail(debugName string, rawValue string) int64 {
+	value, err := strconv.ParseInt(rawValue, 10, 64)
```
same comment.

timothysc commented on test/e2e/networking_perf.go:
```diff
@@ -0,0 +1,249 @@
+
+// This is an opinionated utility method, but could be used with any test that wants to spin N pods on a node.
+// TODO, If we incorporate it in networking.go also, then move it into Utils proper.
+func CreateServiceWithPods(f *Framework, contPort int, svcPort int, appName string, container api.Container, count int) {
```
lower case, or push more generic into utils.

bprashanth commented on test/e2e/networking_perf.go:
```diff
@@ -0,0 +1,249 @@
+	return value
+}
+
+// This is an opinionated utility method, but could be used with any test that wants to spin N pods on a node.
```
I think this would be nicer as an iperf runner. Something that encapsulates all the logic of running iperf that I can invoke from a non-e2e go program as well. It should take just the params for running iperf as struct members, the svc name/ip whatever other runtime args as method args, and look something like:
```
type IPerfRunner struct {
 // iperf args
}
func (iperf *IPerfRunner) Run(port, svcName) IPerfResults {}
```

Then anyone can extend this bench, or add your iperf stuff to other tests.

bprashanth commented on test/e2e/networking_perf.go:
```diff
@@ -0,0 +1,249 @@
+)
+
+const (
+	// This is a magic number, don't rely on it as a well defined constant.
```
please elaborate on what exactly it is, what units, where it's used?

bprashanth commented on test/e2e/networking_perf.go:
```diff
@@ -0,0 +1,249 @@
+	gceBandwidthBitsEstimate = int64(30000000000)
+)
+
+// Declared as Flakey since it has not been proven to run in parallel on small nodes or slow networks in CI
```
s/Flakey/Feature

bprashanth commented on test/e2e/networking_perf.go:
```diff
@@ -0,0 +1,249 @@
+				Name:  "iperf-client",
+				Image: "mattgruter/iperf",
+				Args: []string{
+					"-c", "service-for-" + appName, "-p", "8002", "--reportstyle", "C",
```
please include a comment with the exact iperf command being run including the svc ip/name, for clarity.

bprashanth commented on test/e2e/networking_perf.go:
```diff
@@ -0,0 +1,249 @@
+		bandwidthMap := make(map[string]int64)
+
+		// Calculate expected number of clients based on total nodes.  ITs own function since this
+		// logic may change over time at scale.
```
comment says "ITs own function", meaning?

bprashanth commented on test/e2e/networking_perf.go:
```diff
@@ -0,0 +1,249 @@
+	servPort      int64
+	id            string
+	interval      string
+	transferBits  int64
```
please either comment what eg transferBits represents or include a link to iperf man page if it's the exact same as iperf

jayunit100 commented:
I've updated this to use the framework in #23771 .I like  @bprashanth 's idea about the IPerfRunner externalization... I wonder if that should live in kubernetes? Or somewhere else ?

bprashanth commented:
I'd rather do it right, now, unless it's a prohibitively large refactor.
Less preferred option is to do it in a follow up pr.
Lest preferred is to move it out as a godep (seems to small to incur the overhead).

jayunit100 commented:
gotcha.  i dont think its prohibitively complex to move it out.  im going to test it for stability as is after the rebase .

jayunit100 commented:
I just rebased and am validating this now against the updated cluster framework and so on...

fejta commented on test/e2e/networking_perf.go:
```diff
@@ -0,0 +1,202 @@
+)
+
+// Declared as Flakey since it has not been proven to run in parallel on small nodes or slow networks in CI
+var _ = framework.KubeDescribe("Networking IPerf [Experimental] [Feature:Network-Performance] ", func() {
```
@kubernetes/sig-testing is this how we want to use the Feature: tag?

jayunit100 replied:
You're right ( i think, from what im implying from this comment :) )... In this case, I think maybe based on the semantics, `Feature` could be removed .

timothysc replied:
I would prefer to have a set of [Performance] tags that we could leverage.

jayunit100 commented on test/e2e/framework/framework.go:
```diff
@@ -381,6 +382,64 @@ func (f *Framework) ReadFileViaContainer(podName, containerName string, path str
+// - network throughput on all nodes
+// - soak logs on all nodes.
+// ... and so on
+func (f *Framework) CreateServiceWithPods(contPort int, svcPort int, appName string, podSpec func(n api.Node) api.PodSpec, count int) {
```
@timothysc this function will be reused in #24427

jayunit100 commented:
k8s-bot test this issue: #24507

jayunit100 commented:
@fejta @bprashanth  Let me know if any other thoughts.   I'd like to start iterating on this and look into session affinity etc for scaling servers.

fejta commented:
How long is this test expected to run?

jayunit100 commented:
@fejta generally in under a minute.  
- the actuall throughput measurement is about 10 seconds long.  
- But pulling the container can sometimes be faster/slower.    

We can move the `[Slow]` annotation out in a separate PR after its baked in for a few days.

jayunit100 commented:
@k8s-bot test this issue: #24605

timothysc commented on test/e2e/framework/framework.go:
```diff
@@ -381,6 +382,79 @@ func (f *Framework) ReadFileViaContainer(podName, containerName string, path str
+// CreateServiceForSimpleAppWithPods is a convenience wrapper to create a service and its matching pods all at once.
+func (f *Framework) CreateServiceForSimpleAppWithPods(contPort int, svcPort int, appName string, podSpec func(n api.Node) api.PodSpec, count int, block bool) (error, *api.Service) {
+	var err error = nil
+	theService := f.CreateServiceForSimpleApp(contPort, svcPort, appName)
```
no errors it's all going to work swimmingly?

jayunit100 replied:
It should fail the test if any error see below

timothysc commented on test/e2e/framework/framework.go:
```diff
@@ -381,6 +382,79 @@ func (f *Framework) ReadFileViaContainer(podName, containerName string, path str
+			Selector: serviceSelector,
+		},
+	})
+	ExpectNoError(err)
```
So bail on failure.

jayunit100 replied:
Yeah this is a test failure - the service api call fails,  we expect no error there

timothysc commented on test/e2e/framework/framework.go:
```diff
@@ -381,6 +382,79 @@ func (f *Framework) ReadFileViaContainer(podName, containerName string, path str
+// CreateServiceForSimpleApp returns a service that selects/exposes pods (send -1 ports if no exposure needed) with an app label.
+func (f *Framework) CreateServiceForSimpleApp(contPort, svcPort int, appName string) *api.Service {
+	if appName == "" {
+		panic(fmt.Sprintf("no app name provided"))
```
panic?

jayunit100 replied:
Yeah,since It's not an actual failure, but a bug in the test, it's irrelevant to ginkgo , right?

timothysc replied:
b/c it'a a utility function it should probably pass an error back to be handled.

jayunit100 replied:
+0 I like the declarative test style for readability etc but theoretically your right that side effects should be minimized in lowers functions.

  We have both conventions unfortunately in the code base right now...

Dealers choice let me know, I'm ok either way...

timothysc commented on test/e2e/networking_perf.go:
```diff
@@ -0,0 +1,155 @@
+		)
+
+		if err != nil {
+			framework.Failf("Fatal error waiting for iperf server endpoint : %v", err)
```
does this exit the test here?

jayunit100 replied:
yeah its a ginkgo panic on the spec

jayunit100 commented:
@bprashant PTAL

jayunit100 commented on test/e2e/networking_perf.go:
```diff
@@ -0,0 +1,155 @@
+)
+
+// Declared as Flakey since it has not been proven to run in parallel on small nodes or slow networks in CI
+var _ = framework.KubeDescribe("Networking IPerf [Experimental] [Slow]", func() {
```
Rather than ad hoc tags I can update tags once the strongly typed ones are merged: these tags give us the bare minimum to avoid cluttering the CI

bprashanth replied:
I thought we decided to put this in "feature"?

jayunit100 replied:
We did, but the `Feature` vs `Quirk` thing is being currently defined via sig testing, i figured in the audit that follows we can update all the Feature vs Quirk tags based on the final semantics that get hammered out  (should be within a couple of days)

jayunit100 commented:
@k8s-bot test this please issue: #IGNORE

jayunit100 commented:
( i think travis is flaking again )

bprashanth commented:
@girishkalele maybe there's something between this, netperf tester and what you're trying to do that we can collaborate on?

jayunit100 commented:
Good idea, for us it would be great if  we could iterate against this as its battle tested (I've run it at 300 nodes) and pretty modular.

But if the preference is for me to close this PR that is fine to.

Let me know.

bprashanth commented:
I think this is fine for a first cut, I was thinking more along the lines of how it will evolve. Will iperf be enough for everything? if so, great, if not, should we start off with whatever we think we'll need.

bprashanth commented on test/e2e/framework/framework.go:
```diff
@@ -381,6 +382,79 @@ func (f *Framework) ReadFileViaContainer(podName, containerName string, path str
+		"app": appName + "-pod",
+	}
+	for i, node := range nodes.Items {
+		// Create a MAX of N nodes.
```
create a "max of n nodes"?

bprashanth commented on test/e2e/framework/framework.go:
```diff
@@ -527,6 +601,8 @@ ReturnPodsSoFar:
 	for _, pod := range unfilteredPods {
 		if !(passesPhasesFilter(pod, p.ValidPhases) && passesPodNameFilter(pod, p.PodName)) {
+			// This can be used for debugging but otherwise is very verbose.
+			// Logf("phase or name not matched [[[ %v ]]] : %v %v ", pod, p.ValidPhases, p.PodName)
```
just delete it? or add a slimmer version

bprashanth commented on test/e2e/framework/util.go:
```diff
@@ -2230,6 +2230,20 @@ func ListSchedulableNodesOrDie(c *client.Client) *api.NodeList {
 	return nodes
 }
 
+func ListReadyNodes(f *Framework) (nodes *api.NodeList, err error) {
```
this seems like a common enough thing that we would already have a function that does it, if you've already checked then don't mind this comment

jayunit100 replied:
the Unschedulable part was in networking but now looks like it found its way into utils i guess, so no longer needed.

bprashanth commented on test/e2e/networking_perf.go:
```diff
@@ -0,0 +1,155 @@
+
+const (
+	//empirically derived as a baseline for expectations from running this test using kube-up.sh.
+	gceBandwidthBitsEstimate = int64(30000000000)
```
is this specific to gce? if so how do I set it for any other CP?

jayunit100 replied:
this really is only needed while we bake in the test to see horrible anomolies (its pretty good for that).

bprashanth commented on test/e2e/networking_perf.go:
```diff
@@ -0,0 +1,155 @@
+const (
+	//empirically derived as a baseline for expectations from running this test using kube-up.sh.
+	gceBandwidthBitsEstimate = int64(30000000000)
+	smallClusterTimeout      = 200 * time.Second
```
please add a comment explaining what a "small" cluster is, and what this timeout indicates

bprashanth commented on test/e2e/networking_perf.go:
```diff
@@ -0,0 +1,155 @@
+			return int(math.Min(float64(len(nodes.Items)), float64(numClient)))
+		}()
+
+		//1000 node clusters get an extra 100 seconds to finish.
```
Can you just make this a const ?

jayunit100 replied:
the timeout is dynamic, so it cant be constantified...

bprashanth commented:
This looks fine for a first cut as long as we move it to feature so it doesn't block anything, and the nits are addressed

jayunit100 commented:
@bprashanth comments addressed 
- `GetReadyNodes` seem to have snuck into utils at some point :) . so i just deleted List.
- added 4 node cluster / clarification for smallClusterTimeout
- also clarified the maxCount comment and the dynamic 1/10+seconds timeout that you had questions on .

re: iperf vs netperf vs ... they're all functionally equivalent for the use cases we care about i think.

jayunit100 commented:
Hi @bprashanth any other comments?

bprashanth commented:
> re: iperf vs netperf vs ... they're all functionally equivalent for the use cases we care about i think.

In that case we already have a netperf tester? 
https://github.com/kubernetes/contrib/tree/d9063241280cedfaac23a2283f6584e747aa2302/netperf-tester

LGTM

jayunit100 commented:
@bprashant can we fix this label , thanks.!

jayunit100 commented:
And yeah, netperf as a container can be swapped in here. But netperf-tester you linked appears to be measuring network performance for a simple one off scenario .  This test attempts to scale with respect to cluster size and create replication controllers both for servers and clients...

jayunit100 commented:
@k8s-bot test this please ISSUE: #24577
